From d9e8b8b1dca5b1efda0a81d23be9cdd6cb4038c5 Mon Sep 17 00:00:00 2001
From: Peter Fitger <peter.fitger@flir.se>
Date: Tue, 14 Apr 2015 10:56:07 +0200
Subject: [PATCH] Handle running on AC without battery

Signed-off-by: Peter Fitger <peter.fitger@flir.se>
---
 sleepd.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/sleepd.c b/sleepd.c
index d5ec646..d8cda3d 100644
--- a/sleepd.c
+++ b/sleepd.c
@@ -96,6 +96,8 @@ void apm_read(apm_info * ai)
 			     &ai->battery_time,
 			     &units);
 		close(fh);
+		if ((ai->ac_line_status == APM_AC_UNKNOWN) || (ai->battery_percentage < 0))
+			ai->ac_line_status = APM_AC_ONLINE;
 	}
 }
 
-- 
2.1.0

