From c5eeadfe7b90eb6c787db0b9de90a31432722ebe Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ulf=20Palm=C3=A9r?= <ulf.palmer@flir.se>
Date: Mon, 23 Feb 2015 13:39:28 +0100
Subject: [PATCH] Adapt Makefile for Yocto
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Ulf Palmér <ulf.palmer@flir.se>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ce565c7..94f9dbb 100644
--- a/Makefile
+++ b/Makefile
@@ -25,6 +25,6 @@ clean:
 
 
 flashld: flashld.c load_fpga.c common.c common.h bitswap.c  md5.c Makefile
-	${CROSS_COMPILE}gcc $(CFLAGS) flashld.c load_fpga.c common.c bitswap.c md5.c -o flashld 
+	$(CC) flashld.c load_fpga.c common.c bitswap.c md5.c -o flashld 
 #	${CROSS_COMPILE}strip flashld
 
-- 
1.9.1

