From 9cfe2087fd4ebf527e58dd746f7d6619754d45b1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ulf=20Palm=C3=A9r?= <ulf.palmer@flir.se>
Date: Fri, 19 Sep 2014 12:45:32 +0200
Subject: [PATCH 3/3] Fixed LINKLIBS flag
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Ulf Palmér <ulf.palmer@flir.se>
---
 Makefile.incl | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/Makefile.incl b/Makefile.incl
index e20c78e..4dc72d3 100755
--- a/Makefile.incl
+++ b/Makefile.incl
@@ -60,9 +60,6 @@ SHLIBOBJS := $(addprefix $(SHLIBDIR)/, $(OBJECT2))
 # Create dependency files from OBJECTS in a similar manner
 DEPFILES  := $(patsubst %.o,%.d,$(OBJECTS))
 
-# Expand .so link libraries to full .so.x.y.z name
-LINKLIBS := $(foreach lib,$(LINKLIBS),$(shell ls $(lib)* 2>/dev/null | grep .so...))
-
 # Targets for top-level makefiles
 ifdef SUBDIRS
 .PHONY: subdirs $(SUBDIRS) subdirend clean debugclean
@@ -117,7 +114,7 @@ bin:  $(TARGETDIR)/$(BIN)
 
 $(TARGETDIR)/$(BIN): $(TARGETDIR) $(OBJECTS)
 	echo LD $@
-	$(CC) $(INCLUDES)  $(LDFLAGS) -o $(TARGETDIR)/$(BIN) $(OBJECTS) -Xlinker --start-group $(LINKLIBS) $(LINKSTATICLIBS) -Xlinker --end-group -lc -lstdc++ -lpthread -ldl -lrt -lm
+	$(CC) $(INCLUDES)  $(LDFLAGS) -o $(TARGETDIR)/$(BIN) $(OBJECTS) -Xlinker --start-group $(LINKSTATICLIBS) -Xlinker --end-group -lc -lstdc++ -lpthread -ldl -lrt -lm $(LINKLIBS)
 endif
 
 clean::
-- 
1.8.3.2

