From d4c76f1fdaf78c5f1f72b53bde5360962653bc0a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Mathias=20B=C3=A5ge?= <mathias.bage@flir.com>
Date: Thu, 2 Jun 2022 13:55:53 +0200
Subject: [PATCH] hciattach: change qca baud rate to 4000000

---
 tools/hciattach.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/hciattach.c b/tools/hciattach.c
index b7a41e4..e891e2a 100755
--- a/tools/hciattach.c
+++ b/tools/hciattach.c
@@ -1146,7 +1146,7 @@ struct uart_t uart[] = {
 			FLOW_CTL, DISABLE_PM, NULL, ath3k_ps, ath3k_pm  },
 
 	/* QCA ROME */
-        { "qca",    0x0000, 0x0000, HCI_UART_H4, 115200, 115200,
+        { "qca",    0x0000, 0x0000, HCI_UART_H4, 115200, 4000000,
                         FLOW_CTL, DISABLE_PM, NULL, qca, NULL },
 
 
-- 
2.25.1

