From a842f6c532f5886b743b10bff81710d3bf0444d5 Mon Sep 17 00:00:00 2001
From: David Sernelius <david.sernelius@flir.se>
Date: Thu, 6 Feb 2020 07:04:31 +0100
Subject: [PATCH] cyttsp5: update timer api

---
 drivers/input/touchscreen/cyttsp5_core.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/input/touchscreen/cyttsp5_core.c b/drivers/input/touchscreen/cyttsp5_core.c
index 4cc8b19bf865..4458710e9419 100644
--- a/drivers/input/touchscreen/cyttsp5_core.c
+++ b/drivers/input/touchscreen/cyttsp5_core.c
@@ -3580,9 +3580,9 @@ static void cyttsp5_watchdog_work(struct work_struct *work)
 	cyttsp5_start_wd_timer(cd);
 }
 
-static void cyttsp5_watchdog_timer(unsigned long handle)
+static void cyttsp5_watchdog_timer(struct timer_list *tl)
 {
-	struct cyttsp5_core_data *cd = (struct cyttsp5_core_data *)handle;
+	struct cyttsp5_core_data *cd = from_timer(cd, tl, watchdog_timer);
 
 	if (!cd)
 		return;
@@ -6035,8 +6035,7 @@ int cyttsp5_probe(const struct cyttsp5_bus_ops *ops, struct device *dev,
 	}
 
 	/* Setup watchdog timer */
-	setup_timer(&cd->watchdog_timer, cyttsp5_watchdog_timer,
-			(unsigned long)cd);
+	timer_setup(&cd->watchdog_timer, cyttsp5_watchdog_timer, 0);
 
 	rc = cyttsp5_setup_irq_gpio(cd);
 	if (rc < 0) {
-- 
2.17.1

