From e43163881079e9826b2f924b6520911a484762b2 Mon Sep 17 00:00:00 2001
From: Klas Malmborg <klas.malmborg@teledyneflir.com>
Date: Tue, 3 May 2022 06:20:51 +0200
Subject: [PATCH] BC-105 Buffer out of sync

Signed-off-by: Klas Malmborg <klas.malmborg@teledyneflir.com>
---
 libs/v4l2_core/gstimxv4l2.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libs/v4l2_core/gstimxv4l2.c b/libs/v4l2_core/gstimxv4l2.c
index c0b02a7..08d73ce 100644
--- a/libs/v4l2_core/gstimxv4l2.c
+++ b/libs/v4l2_core/gstimxv4l2.c
@@ -1614,13 +1614,13 @@ gint gst_imx_v4l2_set_buffer_count (gpointer v4l2handle, guint count, guint memo
   memset(&buf_req, 0, sizeof(buf_req));
 
   buf_req.type = handle->type;
-  buf_req.count = MAX_V4L2_BUFFER;
+  buf_req.count = count;
   handle->memory_mode = buf_req.memory = memory_mode;
   if (ioctl(handle->v4l2_fd, VIDIOC_REQBUFS, &buf_req) < 0) {
     GST_ERROR("Request %d buffers failed\n", count);
     return -1;
   }
-  handle->buffer_count = buf_req.count;
+  handle->buffer_count = count;
 
   return 0;
 }
-- 
2.25.1

