diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index e5c3ef9..e263001 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -1893,6 +1893,12 @@ shell_surface_calculate_layer_link (struct shell_surface *shsurf)
 		return &shsurf->shell->fullscreen_layer.view_list;
 	}
 
+        const char *id = weston_desktop_surface_get_app_id(shsurf->desktop_surface);
+        if (id && (strncmp(id, "se.flir.background", 20) == 0)) {
+            weston_log("Setting %s as background\n", id);
+            return &shsurf->shell->background_layer.view_list;
+        }
+
 	/* Move the surface to a normal workspace layer so that surfaces
 	 * which were previously fullscreen or transient are no longer
 	 * rendered on top. */
@@ -2361,6 +2367,9 @@ desktop_surface_added(struct weston_desktop_surface *desktop_surface,
 	struct weston_surface *surface =
 		weston_desktop_surface_get_surface(desktop_surface);
 
+        const char *id = weston_desktop_surface_get_app_id(desktop_surface);
+        weston_log("desktop_surface_added: [%s]\n", id ? id : "<no id>");
+
 	view = weston_desktop_surface_create_view(desktop_surface);
 	if (!view)
 		return;
