diff --git a/compositor/main.c b/compositor/main.c
index 0b48d837..92f61242 100644
--- a/compositor/main.c
+++ b/compositor/main.c
@@ -2431,7 +2431,6 @@ int main(int argc, char *argv[])
 	struct wl_event_loop *loop;
 	int i, fd;
 	char *backend = NULL;
-	char *startup_client = NULL;
 	char *shell = NULL;
 	int32_t xwayland = 0;
 	char *modules = NULL;
@@ -2645,14 +2644,6 @@ int main(int argc, char *argv[])
 	if (argc > 1)
 		goto out;
 
-	/*
-	 * STARTUP CLIENT in weston.ini
-	 */
-	section = weston_config_get_section(config, "flir_startup", NULL, NULL);
-	weston_config_section_get_string(section, "client", &startup_client, NULL);
-	if(startup_client)
-		weston_client_start(wet.compositor, startup_client);
-
 	weston_compositor_wake(wet.compositor);
 
 	wl_display_run(display);
