From 92f86cf236798c221cf64b39d668ae3470103f0b Mon Sep 17 00:00:00 2001
From: David Sernelius <david@tinygiant.se>
Date: Tue, 25 Feb 2020 07:21:53 +0100
Subject: [PATCH] Disables screenshoter permission control

---
 compositor/weston-screenshooter.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/compositor/weston-screenshooter.c b/compositor/weston-screenshooter.c
index a3a8e19..8d36888 100644
--- a/compositor/weston-screenshooter.c
+++ b/compositor/weston-screenshooter.c
@@ -180,12 +180,16 @@ bind_shooter(struct wl_client *client,
 {
 	struct screenshooter *shooter = data;
 	struct wl_resource *resource;
+/*
+ * DISABLE PERMISSION CONTROL TO SEE IF SCREENSHOT IS ALLOWED
+ *
 	bool debug_enabled =
 		weston_compositor_is_debug_protocol_enabled(shooter->ec);
-
+*/
 	resource = wl_resource_create(client,
 				      &weston_screenshooter_interface, 1, id);
 
+/*
 	if (!debug_enabled && !shooter->client) {
 		wl_resource_post_error(resource, WL_DISPLAY_ERROR_INVALID_OBJECT,
 				       "screenshooter failed: permission denied. "\
@@ -196,6 +200,7 @@ bind_shooter(struct wl_client *client,
 				       "screenshooter failed: permission denied.");
 		return;
 	}
+*/
 
 	wl_resource_set_implementation(resource, &screenshooter_implementation,
 				       data, NULL);
-- 
2.7.4

