Index: avahi-0.6.31/avahi-core/announce.c
===================================================================
--- avahi-0.6.31.orig/avahi-core/announce.c
+++ avahi-0.6.31/avahi-core/announce.c
@@ -44,6 +44,12 @@ static void remove_announcer(AvahiServer
     AVAHI_LLIST_REMOVE(AvahiAnnouncer, by_interface, a->interface->announcers, a);
     AVAHI_LLIST_REMOVE(AvahiAnnouncer, by_entry, a->entry->announcers, a);
 
+    if (a->state == AVAHI_PROBING && a->entry->group) {
+        assert(a->entry->group->n_probing);
+	a->entry->group->n_probing--;
+	avahi_s_entry_group_check_probed(a->entry->group, 1);
+    }
+
     avahi_free(a);
 }
 
