From 22dc05403a1532af3f727e6a54176b5fb6d06a3f Mon Sep 17 00:00:00 2001
From: Erik Jansson <Erik.Jansson@teledyneflir.com>
Date: Mon, 7 Mar 2022 08:50:22 +0100
Subject: [PATCH] Add app_id to waylandsink (gst)

---
 ext/wayland/wlwindow.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/ext/wayland/wlwindow.c b/ext/wayland/wlwindow.c
index 71a0288a1..1ea88e8dd 100644
--- a/ext/wayland/wlwindow.c
+++ b/ext/wayland/wlwindow.c
@@ -507,6 +507,8 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
 
     gst_wl_window_ensure_fullscreen (window, fullscreen);
 
+    xdg_toplevel_set_app_id(window->xdg_toplevel, "waylandsink");
+
     /* Finally, commit the xdg_surface state as toplevel */
     window->configured = FALSE;
     wl_surface_commit (window->area_surface);
