From 4cfdd32015080860ae37132d9e8a2c9ab57aaaf8 Mon Sep 17 00:00:00 2001
From: Klas Malmborg <klas.malmborg@teledyneflir.com>
Date: Wed, 5 Apr 2023 10:23:48 +0200
Subject: [PATCH] BC-403 RTSP tunneling causes crash

Signed-off-by: Klas Malmborg <klas.malmborg@teledyneflir.com>
---
 gst/rtsp-server/rtsp-client.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/gst/rtsp-server/rtsp-client.c b/gst/rtsp-server/rtsp-client.c
index 077e1b1..ad3b753 100644
--- a/gst/rtsp-server/rtsp-client.c
+++ b/gst/rtsp-server/rtsp-client.c
@@ -5275,12 +5275,15 @@ gst_rtsp_client_attach (GstRTSPClient * client, GMainContext * context)
 
   gst_rtsp_watch_set_send_backlog (priv->watch, 0, WATCH_BACKLOG_SIZE);
 
+  /* Take the lock before attaching the client watch, so that the clent thread
+   * cannot access the control channel timer until it's properly in place */
+  g_mutex_lock (&priv->lock);
+
   GST_INFO ("client %p: attaching to context %p", client, context);
   res = gst_rtsp_watch_attach (priv->watch, context);
 
   /* Setting up a timeout for the RTSP control channel until a session
    * is up where it is handling timeouts. */
-  g_mutex_lock (&priv->lock);
 
   /* remove old timeout if any */
   rtsp_ctrl_timeout_remove_unlocked (client->priv);
-- 
2.25.1

