From 7bfa1c532756332087907a2d19d1ab4d09a0141d Mon Sep 17 00:00:00 2001
From: David Sernelius <david.sernelius@flir.se>
Date: Tue, 29 Jan 2019 10:09:36 +0100
Subject: [PATCH] Choose manufacturing mode

---
 include/configs/flir_mfgmode.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/flir_mfgmode.h b/include/configs/flir_mfgmode.h
index 208bbda..3ac3102 100644
--- a/include/configs/flir_mfgmode.h
+++ b/include/configs/flir_mfgmode.h
@@ -5,4 +5,4 @@
  *
  */
 
-#define CONFIG_MFG 0
+#define CONFIG_MFG 1
-- 
2.7.4

