From 45b83ca9f343d4176f69826cc4fd6002ee39c143 Mon Sep 17 00:00:00 2001
From: Felix Hammarstrand <felix.hammarstrand@flir.se>
Date: Wed, 23 Oct 2019 16:58:13 +0200
Subject: [PATCH] fixed duplicate config define warnings

---
 include/configs/ec201.h | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/include/configs/ec201.h b/include/configs/ec201.h
index 146af7e..6ae961d 100644
--- a/include/configs/ec201.h
+++ b/include/configs/ec201.h
@@ -32,7 +32,6 @@
 #define CONFIG_SYS_FSL_USDHC_NUM        1
 #define CONFIG_SYS_FSL_ESDHC_ADDR       0
 
-#define CONFIG_CMD_FAT
 
 /* Using ULP WDOG for reset */
 #define WDOG_BASE_ADDR			WDG1_RBASE
@@ -57,8 +56,6 @@
 #define CONFIG_BAUDRATE			115200
 
 #undef CONFIG_CMD_IMLS
-#define CONFIG_SYS_LONGHELP
-#define CONFIG_AUTO_COMPLETE
 
 #define CONFIG_SYS_CACHELINE_SIZE      64
 
@@ -71,7 +68,6 @@
 #define CONFIG_SYS_BARGSIZE CONFIG_SYS_CBSIZE
 #define CONFIG_SYS_PBSIZE (CONFIG_SYS_CBSIZE + sizeof(CONFIG_SYS_PROMPT) + 16)
 
-#define CONFIG_CMDLINE_EDITING
 #define CONFIG_STACKSIZE		SZ_8K
 
 /* Physical Memory Map */
@@ -82,7 +78,6 @@
 #define PHYS_SDRAM_SIZE			SZ_512M
 #define CONFIG_SYS_MEMTEST_START	PHYS_SDRAM
 #define CONFIG_SYS_SDRAM_BASE		PHYS_SDRAM
-#define CONFIG_CMD_BOOTZ
 
 #define CONFIG_LOADADDR             0x60800000
 
@@ -121,7 +116,6 @@
 #define CONFIG_VIDEO_BMP_GZIP
 #define CONFIG_SYS_VIDEO_LOGO_MAX_SIZE 1024*1024
 #endif
-
 #define CONFIG_OF_BOARD_SETUP
 
 /* Override imx standard behavior when loading u-boot from usb */
@@ -160,7 +154,6 @@
 
 #else /* CONFIG_MFG */
 
-#define CONFIG_ENV_IS_IN_MMC
 #define CONFIG_SYS_MMC_ENV_DEV		0		/* emmc0 */
 #define CONFIG_ENV_SIZE			SZ_16K
 #define CONFIG_ENV_OFFSET		(1792 * 1024)	/* 256kB below 2MiB */
