From 46b3e81f17e214d6eac750414ff46686c3d636fa Mon Sep 17 00:00:00 2001
From: Erik Bengtsson <erik.bengtsson@flir.se>
Date: Mon, 31 Aug 2020 10:41:39 +0200
Subject: [PATCH] lm3642: Turn off torch at shutdown.

---
 drivers/leds/leds-lm3642.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/leds/leds-lm3642.c b/drivers/leds/leds-lm3642.c
index 6235fb1d27a7..69dbb4665fe9 100644
--- a/drivers/leds/leds-lm3642.c
+++ b/drivers/leds/leds-lm3642.c
@@ -477,6 +477,9 @@ static int lm3642_probe(struct i2c_client *client,
 		goto err_create_indicator_file;
 	}
 
+	//Always start with torch off
+	lm3642_control(chip, 0, MODES_TORCH);
+
 	dev_info(&client->dev, "LM3642 is initialized\n");
 	return 0;
 
@@ -488,14 +491,21 @@ static int lm3642_probe(struct i2c_client *client,
 	return err;
 }
 
-static int lm3642_remove(struct i2c_client *client)
+static void lm3642_shutdown(struct i2c_client *client)
 {
 	struct lm3642_chip_data *chip = i2c_get_clientdata(client);
 
+	lm3642_control(chip, 0, MODES_TORCH);
+
 	led_classdev_unregister(&chip->cdev_indicator);
 	led_classdev_unregister(&chip->cdev_torch);
 	led_classdev_unregister(&chip->cdev_flash);
 	regmap_write(chip->regmap, REG_ENABLE, 0);
+}
+
+static int lm3642_remove(struct i2c_client *client)
+{
+	lm3642_shutdown(client);
 	return 0;
 }
 
@@ -513,6 +523,7 @@ static struct i2c_driver lm3642_i2c_driver = {
 		   },
 	.probe = lm3642_probe,
 	.remove = lm3642_remove,
+	.shutdown = lm3642_shutdown,
 	.id_table = lm3642_id,
 };
 
-- 
2.17.1

