From 8d3e7e858938bbbe9791315f81fdaea56428ae0f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ulf=20Palm=C3=A9r?= <ulf.palmer@flir.se>
Date: Wed, 15 Jan 2020 08:46:15 +0100
Subject: [PATCH] Add splash logo to env

---
 include/configs/ec201.h | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/include/configs/ec201.h b/include/configs/ec201.h
index e1af049..729a822 100644
--- a/include/configs/ec201.h
+++ b/include/configs/ec201.h
@@ -367,7 +367,11 @@
 				"bootaux ${m4_addr};" \
 			"fi;\0"
 
-
+#define CONFIG_SPLASH_IMAGE_ENV \
+	"splashimage=0x67000000\0" \
+	"splashsource=mmc_fs\0" \
+	"splashfile=/boot/bootlogo.bmp.gz\0" \
+	"panel=TRULY-VGA-SHERLOCK\0"
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	CONFIG_DEFAULT_NETWORK_SETTINGS \
@@ -387,6 +391,7 @@
 	CONFIG_SELECT_BOOT_PARTITION_ENV \
 	CONFIG_MMC_BOOT_FLIR_ENV\
 	CONFIG_MMC_BOOT_LEGACY_ENV \
+	CONFIG_SPLASH_IMAGE_ENV \
 	""	/* end line */
 
 
-- 
1.9.1

