From 3bad313ab711427ea4998678e05604e426741326 Mon Sep 17 00:00:00 2001
From: Felix Hammarstrand <felix.hammarstrand@flir.se>
Date: Tue, 10 Sep 2019 10:37:14 +0200
Subject: [PATCH] sherlock display: fixed pixel clock

---
 board/flir/ec201/ec201.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/board/flir/ec201/ec201.c b/board/flir/ec201/ec201.c
index 85c7fd8..8477064 100644
--- a/board/flir/ec201/ec201.c
+++ b/board/flir/ec201/ec201.c
@@ -162,7 +162,7 @@ struct display_info_t const displays[] = {{
 		.name			= "TRULY-VGA-SHERLOCK",
 		.xres           = 640,
 		.yres           = 480,
-		.pixclock       = 468984,
+		.pixclock       = 46894,
 		.left_margin    = 15,
 		.right_margin   = 26,
 		.upper_margin   = 9,
